samples: bluetooth: mesh: Shell client sample

HaavardRei opened the pull request:
Adds a mesh shell client that supports sending scheduler, scene, time-
and light control client model messages.

Modifies the existing light_ctrl sample to support corresponding
model servers.

Signed-off-by: Håvard Reierstad <havardreierstad@gmail.com>

HaavardRei commented:
Created new PR due to naming conflicts, link to old PR here:
https://github.com/HaavardRei/sdk-nrf/pull/1

Andrewpini commented:
Please do another rebase on top of your fork main. This should remove alterations that is added by other people than you

Andrewpini commented on samples/bluetooth/mesh/light_ctrl/src/model_handler.c:
```diff
@@ -80,7 +85,7 @@ static void start_new_light_trans(const struct bt_mesh_lightness_set *set,
 {
 	uint32_t step_cnt = abs(set->lvl - ctx->current_lvl) / PWM_SIZE_STEP;
-	uint32_t time = set->transition ? set->transition->time : 0;
+	uint32_t time =	 set->transition ? set->transition->time : 0;
```
Remove redundant space
```suggestion
	uint32_t time = set->transition ? set->transition->time : 0;
```

Andrewpini commented on samples/bluetooth/mesh/light_ctrl/prj.conf:
```diff
@@ -48,7 +48,13 @@ CONFIG_BT_MESH_GATT_PROXY=y
 CONFIG_BT_MESH_LIGHT_CTRL_SRV=y
-CONFIG_BT_MESH_LIGHT_CTRL_SRV_TIME_ON=3
+CONFIG_BT_MESH_LIGHT_CTRL_SRV_TIME_ON=6
 CONFIG_BT_MESH_LIGHT_CTRL_SRV_TIME_PROLONG=3
-CONFIG_BT_MESH_LIGHT_CTRL_SRV_RESUME_DELAY=30
+CONFIG_BT_MESH_LIGHT_CTRL_SRV_RESUME_DELAY=0
```
Can you elaborate on why you have changed these parameters. Is it essential to your feature?

HaavardRei replied:
I added a comment about this in the previous PR, but copy paste:

I chose to change this to six, because when recalling the scene specified in toggle_scene.py (with five seconds transition time), the light will begin to fade before reaching max level. I guess this is due to the LC state machine being turned on at the beginning of the transition, then taking over before the transition completes.

The resume_delay was set to zero for testing purposes, but should probably be included if we want the LC state machine to be automatically reenabled

HaavardRei replied:
When having 30s as resume_delay it gives a bit wonky behavior when scheduling periodic on/off tasks, such as in toggle_onoff.py. The timer kicks in 30 seconds after the LC state machine gets disabled, which hijacks the lights when a scheduled task is fired. The state machine gets disabled on the next scheduled action, but then comes back 30s later... Maybe there should be some functionality that prevents this from happening/postpones the resumption when other actions are fired?

Andrewpini commented on samples/bluetooth/mesh/light_ctrl/prj.conf:
```diff
@@ -48,7 +48,13 @@ CONFIG_BT_MESH_GATT_PROXY=y
+# Bluetooth debug
+CONFIG_BT_DEBUG_LOG=y
+CONFIG_BT_MESH_DEBUG=y
+CONFIG_BT_MESH_DEBUG_MODEL=y
```
While I think it could be a good idea to provide debug options to the users of the samples, this is not something that is done for the other samples we provide ATM. I think it would be best to keep them aligned to each other for now. So you can remove it for now.

If we wanted to provide the options in the project config, I think it should look like this:
```suggestion
# Bluetooth debug
# CONFIG_BT_DEBUG_LOG=y
# CONFIG_BT_MESH_DEBUG=y
# CONFIG_BT_MESH_DEBUG_MODEL=y
```
This would ensure that debug is not enabled by default, but provides a user/customer an easy way to enable them at will.

HaavardRei replied:
Commented them out in the second commit for them to easily be reenabled, but can fully remove them in a final version.

Andrewpini commented on samples/bluetooth/mesh/light_ctrl/src/model_handler.c:
```diff
@@ -185,15 +193,14 @@ static const struct bt_mesh_comp comp = {
 void model_handler_start(void)
 {
 	int err;
-
```
Nitpick: This newline should be preserved

Andrewpini commented on samples/bluetooth/mesh/light_ctrl/src/model_handler.c:
```diff
@@ -21,6 +21,11 @@ struct lightness_ctx {
+	struct bt_mesh_time_srv time_srv;
+	struct bt_mesh_scheduler_srv scheduler_srv;
+};
+
```
I see what you are trying to do whit this context, but I'm wondering if it is neccesary. Of course it implies that this scheduler and time server belongs together, but that is also implied in the initialization of the scheduler.

I would suggest to keep it simple with somehing like this:

```
static struct bt_mesh_time_srv time_srv = BT_MESH_TIME_SRV_INIT(NULL);
static struct bt_mesh_scheduler_srv scheduler_srv = BT_MESH_SCHEDULER_SRV_INIT(NULL, time_srv);
``` 

It is not a big deal, but introducing a separate struct and an extra nesting level feels somewhat redundant to me in this particular case.

Andrewpini commented on samples/bluetooth/mesh/light_ctrl/README.rst:
```diff
@@ -76,18 +85,32 @@ The following table shows the mesh light fixture composition data for this sampl
 * The three models in the second element are the product of a single instance of the Light Lightness Control (LC) Server.
   The Light LC Server controls the Light Lightness Server in the first element, deciding on parameters such as fade time, lighting levels for different states, and inactivity timing.
-  In this sample, the Light LC Server is enabled by default at first boot.
+  In this sample, the Light LC Server is enabled by default at first boot. <---- Not true for current version.
```
If you have observed this and verified it you can just remove the entire line.
In that case, nice catch 👍

HaavardRei replied:
I think it depends on the resume_delay config option, will investigate!

HaavardRei replied:
Huh, it's enabled when first flashing after erasing, but when I provisioned it and configured the appkeys there seems to be something that prevents it from getting enabled (at least in the model_handler_start function, it can be enabled from shell). 
![image](https://user-images.githubusercontent.com/91412738/177963854-4ff5c2b3-94b1-41d1-8598-d10ccbdf0bb6.png)

HaavardRei commented on samples/bluetooth/mesh/shell_client/README.rst:
```diff
@@ -0,0 +1,193 @@
+* A python script :file:`store_scenes.py` that can be used to generate scenes to be used in the point above.
+
+
+FEM support
```
This might not be relevant for the sample/should be tested before including

Andrewpini replied:
To be honest I'm not quite sure. Just leave it as is for now.

Andrewpini commented on samples/bluetooth/mesh/shell_client/prj.conf:
```diff
@@ -0,0 +1,60 @@
+CONFIG_BT_MESH_SCENE_CLI=y
+CONFIG_BT_MESH_TIME_CLI=y
+CONFIG_BT_MESH_LIGHT_CTRL_CLI=y
+CONFIG_BT_MESH_SHELL=y
```
Nitpick:

```suggestion
# Shell support
CONFIG_BT_MESH_SHELL=y

# Bluetooth mesh models
CONFIG_BT_MESH_SCHEDULER_CLI=y
CONFIG_BT_MESH_SCENE_CLI=y
CONFIG_BT_MESH_TIME_CLI=y
CONFIG_BT_MESH_LIGHT_CTRL_CLI=y
```

Andrewpini commented on samples/bluetooth/mesh/shell_client/src/model_handler.c:
```diff
@@ -0,0 +1,96 @@
+#include <dk_buttons_and_leds.h>
+#include "model_handler.h"
+
+
+
```
Remove redundant newlines

Andrewpini commented on samples/bluetooth/mesh/shell_client/src/model_handler.c:
```diff
@@ -0,0 +1,96 @@
+/* Set up a repeating delayed work to blink the DK's LEDs when attention is
+ * requested.
+ */
+
```
Remove redundant newline

Andrewpini commented on samples/bluetooth/mesh/shell_client/store_scenes.py:
```diff
@@ -0,0 +1,22 @@
+
+ser = serial.Serial()
+ser.baudrate = 115200
+ser.port = 'COM6'        # Serial port of client DK
```
Maybe this can be done as a input from cmd line? The COM port will be different for each induvidual setup.
Applies to all your scripts.
https://www.w3schools.com/python/ref_func_input.asp

Andrewpini commented on samples/bluetooth/mesh/shell_client/README.rst:
```diff
@@ -0,0 +1,193 @@
+.. _bluetooth_mesh_shell_client:
```
In general I think this documentation looks fine. Since the capabillities of a client that is hooked up to mesh shell exeeds the mentioned usecases that is described here, I'm not 100% sure what will be the final result of this sample when it is going to be merged to NCS. But it will be close to what you have in your current version. I just want you to keep that in mind before spending too much time ironing out the last details of the docs here.

Andrewpini reviewed (commented):
In general this looks very good. I haven't tested the feature myself yet, but if I assume that it is functional I think you are almost ready to run a demo of it 👍

Andrewpini reviewed (commented):
Nice! This looks good to me (At least for now). Can you sqaush the two commits into a single one? It is redundant to keep them as separate commits. If you for some personal reason want to keep them separated you can postpone squashing them until later.

HaavardRei committed:
samples: bluetooth: mesh: shell client sample

Adds a mesh shell client that supports sending scheduler-, scene-, time-
and light control client model messages.

Modifies the existing light_ctrl sample to support corresponding
model servers.

Signed-off-by: Håvard Reierstad <havardreierstad@gmail.com>
